{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "12c3c518",
   "metadata": {},
   "source": [
    "# Human Gene Length Database"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "987cb1ec",
   "metadata": {},
   "source": [
    "## Downloading Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a82eb5b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Downloads and unzips basic annotations\n",
    "!wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_36/gencode.v36.basic.annotation.gtf.gz\n",
    "!gunzip gencode.v36.basic.annotation.gtf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c2ebe43",
   "metadata": {},
   "source": [
    "## Subsetting Genes, Transcripts, and Exons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35e760b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subsets genes, transcripts, and exons\n",
    "!grep -P \"\\tgene\\t\" gencode.v36.basic.annotation.gtf > genes.gtf\n",
    "!grep -P \"\\ttranscript\\t\" gencode.v36.basic.annotation.gtf > transcripts.gtf\n",
    "!grep -P \"\\texon\\t\" gencode.v36.basic.annotation.gtf > exons.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "459f42e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# In the GTF format, columns are separated by tabs\n",
    "# and the 9th column has several fields separated by \";\"\n",
    "# Genes, transcripts, and exons have different sets of \n",
    "# fields in the 9th column, so we need to process them\n",
    "# individually\n",
    "\n",
    "# Splits gene columns into two files for loading and re-merging\n",
    "!cut -f1-8 genes.gtf > genes_basic.gtf\n",
    "!cut -f9 genes.gtf > genes_extra.gtf\n",
    "# Splits transcript columns into two files for loading and re-merging\n",
    "!cut -f1-8 transcripts.gtf > transcripts_basic.gtf\n",
    "!cut -f9 transcripts.gtf > transcripts_extra.gtf\n",
    "# Splits exon columns into two files for loading and re-merging\n",
    "!cut -f1-8 exons.gtf > exons_basic.gtf\n",
    "!cut -f9 exons.gtf > exons_extra.gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95dd8015",
   "metadata": {},
   "source": [
    "## Loading Data into Pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fe0a2d51",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def gen_df(feature):\n",
    "    # Loads columns 1-8\n",
    "    basic = pd.read_csv(f'./{feature}_basic.gtf', sep='\\t', header=None, names=['chr', 'program', 'feature', 'start', 'end', 'score', 'strand', 'frame'])\n",
    "    # Loads column 9, split into its fields\n",
    "    extra = pd.read_csv(f'./{feature}_extra.gtf', header=None)\n",
    "    # Builds a dictionary with the existing fields\n",
    "    # and turns it into a DataFrame\n",
    "    extra = extra.map(lambda v: dict([elem.strip().split(' ') for elem in v[:-1].split(';')]))\n",
    "    extra = pd.DataFrame(list(extra.values.squeeze()))\n",
    "    # Removes extra double-quotes from string fields\n",
    "    extra = extra.map(lambda s: s[1:-1] if isinstance(s, str) and s[0] == '\"' else s)\n",
    "    # Merge them together\n",
    "    df = pd.concat([basic, extra], axis=1).drop(columns=['score', 'frame'])\n",
    "    # Compute length of the feature\n",
    "    df['length'] = df['end'] - df['start'] + 1\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "74b59a6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transforms genes, transcripts, and exons files into DFs\n",
    "genes = gen_df('genes')\n",
    "transcripts = gen_df('transcripts')\n",
    "exons = gen_df('exons')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d8cecf4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>program</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>tag</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>29554</td>\n",
       "      <td>31109</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>2</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>ncRNA_host</td>\n",
       "      <td>1556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:35294</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60655</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60656</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7479</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60657</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7427</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60658</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60659</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>3</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60660 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr  program feature  start    end strand            gene_id  \\\n",
       "0      chr1   HAVANA    gene  11869  14409      +  ENSG00000223972.5   \n",
       "1      chr1   HAVANA    gene  14404  29570      -  ENSG00000227232.5   \n",
       "2      chr1  ENSEMBL    gene  17369  17436      -  ENSG00000278267.1   \n",
       "3      chr1   HAVANA    gene  29554  31109      +  ENSG00000243485.5   \n",
       "4      chr1  ENSEMBL    gene  30366  30503      +  ENSG00000284332.1   \n",
       "...     ...      ...     ...    ...    ...    ...                ...   \n",
       "60655  chrM  ENSEMBL    gene  14149  14673      -  ENSG00000198695.2   \n",
       "60656  chrM  ENSEMBL    gene  14674  14742      -  ENSG00000210194.1   \n",
       "60657  chrM  ENSEMBL    gene  14747  15887      +  ENSG00000198727.2   \n",
       "60658  chrM  ENSEMBL    gene  15888  15953      +  ENSG00000210195.2   \n",
       "60659  chrM  ENSEMBL    gene  15956  16023      -  ENSG00000210196.2   \n",
       "\n",
       "                                gene_type    gene_name level     hgnc_id  \\\n",
       "0      transcribed_unprocessed_pseudogene      DDX11L1     2  HGNC:37102   \n",
       "1                  unprocessed_pseudogene       WASH7P     2  HGNC:38034   \n",
       "2                                   miRNA    MIR6859-1     3  HGNC:50039   \n",
       "3                                  lncRNA  MIR1302-2HG     2  HGNC:52482   \n",
       "4                                   miRNA    MIR1302-2     3  HGNC:35294   \n",
       "...                                   ...          ...   ...         ...   \n",
       "60655                      protein_coding       MT-ND6     3   HGNC:7462   \n",
       "60656                             Mt_tRNA        MT-TE     3   HGNC:7479   \n",
       "60657                      protein_coding       MT-CYB     3   HGNC:7427   \n",
       "60658                             Mt_tRNA        MT-TT     3   HGNC:7499   \n",
       "60659                             Mt_tRNA        MT-TP     3   HGNC:7494   \n",
       "\n",
       "                havana_gene         tag  length  \n",
       "0      OTTHUMG00000000961.2         NaN    2541  \n",
       "1      OTTHUMG00000000958.1         NaN   15167  \n",
       "2                       NaN         NaN      68  \n",
       "3      OTTHUMG00000000959.2  ncRNA_host    1556  \n",
       "4                       NaN         NaN     138  \n",
       "...                     ...         ...     ...  \n",
       "60655                   NaN         NaN     525  \n",
       "60656                   NaN         NaN      69  \n",
       "60657                   NaN         NaN    1141  \n",
       "60658                   NaN         NaN      66  \n",
       "60659                   NaN         NaN      68  \n",
       "\n",
       "[60660 rows x 14 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "681f00c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>program</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>11869</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>12010</td>\n",
       "      <td>13670</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14404</td>\n",
       "      <td>29570</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:38034</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>PGO:0000005</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:50039</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>29554</td>\n",
       "      <td>31097</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>HGNC:52482</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110077</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7462</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000354665.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110078</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>ENST00000387459.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7479</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110079</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7427</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000354554.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110080</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110081</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110082 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr  program     feature  start    end strand            gene_id  \\\n",
       "0       chr1   HAVANA  transcript  11869  14409      +  ENSG00000223972.5   \n",
       "1       chr1   HAVANA  transcript  12010  13670      +  ENSG00000223972.5   \n",
       "2       chr1   HAVANA  transcript  14404  29570      -  ENSG00000227232.5   \n",
       "3       chr1  ENSEMBL  transcript  17369  17436      -  ENSG00000278267.1   \n",
       "4       chr1   HAVANA  transcript  29554  31097      +  ENSG00000243485.5   \n",
       "...      ...      ...         ...    ...    ...    ...                ...   \n",
       "110077  chrM  ENSEMBL  transcript  14149  14673      -  ENSG00000198695.2   \n",
       "110078  chrM  ENSEMBL  transcript  14674  14742      -  ENSG00000210194.1   \n",
       "110079  chrM  ENSEMBL  transcript  14747  15887      +  ENSG00000198727.2   \n",
       "110080  chrM  ENSEMBL  transcript  15888  15953      +  ENSG00000210195.2   \n",
       "110081  chrM  ENSEMBL  transcript  15956  16023      -  ENSG00000210196.2   \n",
       "\n",
       "            transcript_id                           gene_type    gene_name  \\\n",
       "0       ENST00000456328.2  transcribed_unprocessed_pseudogene      DDX11L1   \n",
       "1       ENST00000450305.2  transcribed_unprocessed_pseudogene      DDX11L1   \n",
       "2       ENST00000488147.1              unprocessed_pseudogene       WASH7P   \n",
       "3       ENST00000619216.1                               miRNA    MIR6859-1   \n",
       "4       ENST00000473358.1                              lncRNA  MIR1302-2HG   \n",
       "...                   ...                                 ...          ...   \n",
       "110077  ENST00000361681.2                      protein_coding       MT-ND6   \n",
       "110078  ENST00000387459.1                             Mt_tRNA        MT-TE   \n",
       "110079  ENST00000361789.2                      protein_coding       MT-CYB   \n",
       "110080  ENST00000387460.2                             Mt_tRNA        MT-TT   \n",
       "110081  ENST00000387461.2                             Mt_tRNA        MT-TP   \n",
       "\n",
       "        ... level transcript_support_level     hgnc_id                 tag  \\\n",
       "0       ...     2                        1  HGNC:37102               basic   \n",
       "1       ...     2                       NA  HGNC:37102               basic   \n",
       "2       ...     2                       NA  HGNC:38034               basic   \n",
       "3       ...     3                       NA  HGNC:50039               basic   \n",
       "4       ...     2                        5  HGNC:52482               basic   \n",
       "...     ...   ...                      ...         ...                 ...   \n",
       "110077  ...     3                       NA   HGNC:7462  appris_principal_1   \n",
       "110078  ...     3                       NA   HGNC:7479               basic   \n",
       "110079  ...     3                       NA   HGNC:7427  appris_principal_1   \n",
       "110080  ...     3                       NA   HGNC:7499               basic   \n",
       "110081  ...     3                       NA   HGNC:7494               basic   \n",
       "\n",
       "                 havana_gene     havana_transcript          ont  \\\n",
       "0       OTTHUMG00000000961.2  OTTHUMT00000362751.1          NaN   \n",
       "1       OTTHUMG00000000961.2  OTTHUMT00000002844.2  PGO:0000019   \n",
       "2       OTTHUMG00000000958.1  OTTHUMT00000002839.1  PGO:0000005   \n",
       "3                        NaN                   NaN          NaN   \n",
       "4       OTTHUMG00000000959.2  OTTHUMT00000002840.1          NaN   \n",
       "...                      ...                   ...          ...   \n",
       "110077                   NaN                   NaN          NaN   \n",
       "110078                   NaN                   NaN          NaN   \n",
       "110079                   NaN                   NaN          NaN   \n",
       "110080                   NaN                   NaN          NaN   \n",
       "110081                   NaN                   NaN          NaN   \n",
       "\n",
       "               protein_id ccdsid length  \n",
       "0                     NaN    NaN   2541  \n",
       "1                     NaN    NaN   1661  \n",
       "2                     NaN    NaN  15167  \n",
       "3                     NaN    NaN     68  \n",
       "4                     NaN    NaN   1544  \n",
       "...                   ...    ...    ...  \n",
       "110077  ENSP00000354665.2    NaN    525  \n",
       "110078                NaN    NaN     69  \n",
       "110079  ENSP00000354554.2    NaN   1141  \n",
       "110080                NaN    NaN     66  \n",
       "110081                NaN    NaN     68  \n",
       "\n",
       "[110082 rows x 22 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ba691ec9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>program</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>...</th>\n",
       "      <th>level</th>\n",
       "      <th>transcript_support_level</th>\n",
       "      <th>hgnc_id</th>\n",
       "      <th>tag</th>\n",
       "      <th>havana_gene</th>\n",
       "      <th>havana_transcript</th>\n",
       "      <th>ont</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>ccdsid</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:37102</td>\n",
       "      <td>basic</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>PGO:0000019</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771718</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7462</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000354665.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771719</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>ENST00000387459.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7479</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771720</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7427</td>\n",
       "      <td>appris_principal_1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ENSP00000354554.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771721</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>+</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7499</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771722</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSEMBL</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>NA</td>\n",
       "      <td>HGNC:7494</td>\n",
       "      <td>basic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>771723 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr  program feature  start    end strand            gene_id  \\\n",
       "0       chr1   HAVANA    exon  11869  12227      +  ENSG00000223972.5   \n",
       "1       chr1   HAVANA    exon  12613  12721      +  ENSG00000223972.5   \n",
       "2       chr1   HAVANA    exon  13221  14409      +  ENSG00000223972.5   \n",
       "3       chr1   HAVANA    exon  12010  12057      +  ENSG00000223972.5   \n",
       "4       chr1   HAVANA    exon  12179  12227      +  ENSG00000223972.5   \n",
       "...      ...      ...     ...    ...    ...    ...                ...   \n",
       "771718  chrM  ENSEMBL    exon  14149  14673      -  ENSG00000198695.2   \n",
       "771719  chrM  ENSEMBL    exon  14674  14742      -  ENSG00000210194.1   \n",
       "771720  chrM  ENSEMBL    exon  14747  15887      +  ENSG00000198727.2   \n",
       "771721  chrM  ENSEMBL    exon  15888  15953      +  ENSG00000210195.2   \n",
       "771722  chrM  ENSEMBL    exon  15956  16023      -  ENSG00000210196.2   \n",
       "\n",
       "            transcript_id                           gene_type gene_name  ...  \\\n",
       "0       ENST00000456328.2  transcribed_unprocessed_pseudogene   DDX11L1  ...   \n",
       "1       ENST00000456328.2  transcribed_unprocessed_pseudogene   DDX11L1  ...   \n",
       "2       ENST00000456328.2  transcribed_unprocessed_pseudogene   DDX11L1  ...   \n",
       "3       ENST00000450305.2  transcribed_unprocessed_pseudogene   DDX11L1  ...   \n",
       "4       ENST00000450305.2  transcribed_unprocessed_pseudogene   DDX11L1  ...   \n",
       "...                   ...                                 ...       ...  ...   \n",
       "771718  ENST00000361681.2                      protein_coding    MT-ND6  ...   \n",
       "771719  ENST00000387459.1                             Mt_tRNA     MT-TE  ...   \n",
       "771720  ENST00000361789.2                      protein_coding    MT-CYB  ...   \n",
       "771721  ENST00000387460.2                             Mt_tRNA     MT-TT  ...   \n",
       "771722  ENST00000387461.2                             Mt_tRNA     MT-TP  ...   \n",
       "\n",
       "       level transcript_support_level     hgnc_id                 tag  \\\n",
       "0          2                        1  HGNC:37102               basic   \n",
       "1          2                        1  HGNC:37102               basic   \n",
       "2          2                        1  HGNC:37102               basic   \n",
       "3          2                       NA  HGNC:37102               basic   \n",
       "4          2                       NA  HGNC:37102               basic   \n",
       "...      ...                      ...         ...                 ...   \n",
       "771718     3                       NA   HGNC:7462  appris_principal_1   \n",
       "771719     3                       NA   HGNC:7479               basic   \n",
       "771720     3                       NA   HGNC:7427  appris_principal_1   \n",
       "771721     3                       NA   HGNC:7499               basic   \n",
       "771722     3                       NA   HGNC:7494               basic   \n",
       "\n",
       "                 havana_gene     havana_transcript          ont  \\\n",
       "0       OTTHUMG00000000961.2  OTTHUMT00000362751.1          NaN   \n",
       "1       OTTHUMG00000000961.2  OTTHUMT00000362751.1          NaN   \n",
       "2       OTTHUMG00000000961.2  OTTHUMT00000362751.1          NaN   \n",
       "3       OTTHUMG00000000961.2  OTTHUMT00000002844.2  PGO:0000019   \n",
       "4       OTTHUMG00000000961.2  OTTHUMT00000002844.2  PGO:0000019   \n",
       "...                      ...                   ...          ...   \n",
       "771718                   NaN                   NaN          NaN   \n",
       "771719                   NaN                   NaN          NaN   \n",
       "771720                   NaN                   NaN          NaN   \n",
       "771721                   NaN                   NaN          NaN   \n",
       "771722                   NaN                   NaN          NaN   \n",
       "\n",
       "               protein_id ccdsid length  \n",
       "0                     NaN    NaN    359  \n",
       "1                     NaN    NaN    109  \n",
       "2                     NaN    NaN   1189  \n",
       "3                     NaN    NaN     48  \n",
       "4                     NaN    NaN     49  \n",
       "...                   ...    ...    ...  \n",
       "771718  ENSP00000354665.2    NaN    525  \n",
       "771719                NaN    NaN     69  \n",
       "771720  ENSP00000354554.2    NaN   1141  \n",
       "771721                NaN    NaN     66  \n",
       "771722                NaN    NaN     68  \n",
       "\n",
       "[771723 rows x 24 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11ff97e1",
   "metadata": {},
   "source": [
    "## Sorting Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "377c1f13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creates a sortable chromosome info\n",
    "exons['chr_sort'] = exons['chr'].apply(lambda v: v[:3]+'0'+v[-1] if v[-1].isnumeric() and len(v) == 4 else (v if v[-1] != 'M' else 'chr_M'))\n",
    "# Sorts genes in the order they appear in chromosomes\n",
    "ordered_genes = (exons\n",
    "                 .groupby(['chr_sort', 'gene_id'])['start']\n",
    "                 .min()\n",
    "                 .reset_index()\n",
    "                 .sort_values(by=['chr_sort', 'start'])\n",
    "                 .reset_index(drop=True))\n",
    "\n",
    "# Creates a sortable gene info\n",
    "ordered_genes['gene_sort'] = list(map(lambda v: f'{(v+1):05d}', list(ordered_genes.index)))\n",
    "exons = exons.merge(ordered_genes[['gene_id', 'gene_sort']], on='gene_id')\n",
    "\n",
    "# Sorts transcripts in the order they appear in genes\n",
    "ordered_tx = (exons\n",
    "              .groupby(['chr_sort', 'gene_sort', 'transcript_id'])['start']\n",
    "              .min()\n",
    "              .reset_index()\n",
    "              .sort_values(by=['chr_sort', 'gene_sort', 'start'])\n",
    "              .reset_index(drop=True))\n",
    "# Creates a sortable tx info\n",
    "ordered_tx['tx_sort'] = list(map(lambda v: f'{(v+1):06d}', list(ordered_tx.index)))\n",
    "exons = exons.merge(ordered_tx[['transcript_id', 'tx_sort']], on='transcript_id')\n",
    "\n",
    "# Sorts exons by chromosome, gene, and transcript\n",
    "exons = (exons\n",
    "         .sort_values(by=['chr_sort', 'gene_sort', 'tx_sort', 'start'])\n",
    "         [['chr', 'gene_id', 'transcript_id', 'exon_id', 'start', 'end', 'length']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ba057406",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>1189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771718</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>ENSE00001434974.2</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771719</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>ENST00000387459.1</td>\n",
       "      <td>ENSE00001544476.1</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771720</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>ENSE00001436074.2</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771721</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>ENSE00001544475.2</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771722</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>ENSE00001544473.2</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>771723 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr            gene_id      transcript_id            exon_id  start  \\\n",
       "0       chr1  ENSG00000223972.5  ENST00000456328.2  ENSE00002234944.1  11869   \n",
       "1       chr1  ENSG00000223972.5  ENST00000456328.2  ENSE00003582793.1  12613   \n",
       "2       chr1  ENSG00000223972.5  ENST00000456328.2  ENSE00002312635.1  13221   \n",
       "3       chr1  ENSG00000223972.5  ENST00000450305.2  ENSE00001948541.1  12010   \n",
       "4       chr1  ENSG00000223972.5  ENST00000450305.2  ENSE00001671638.2  12179   \n",
       "...      ...                ...                ...                ...    ...   \n",
       "771718  chrM  ENSG00000198695.2  ENST00000361681.2  ENSE00001434974.2  14149   \n",
       "771719  chrM  ENSG00000210194.1  ENST00000387459.1  ENSE00001544476.1  14674   \n",
       "771720  chrM  ENSG00000198727.2  ENST00000361789.2  ENSE00001436074.2  14747   \n",
       "771721  chrM  ENSG00000210195.2  ENST00000387460.2  ENSE00001544475.2  15888   \n",
       "771722  chrM  ENSG00000210196.2  ENST00000387461.2  ENSE00001544473.2  15956   \n",
       "\n",
       "          end  length  \n",
       "0       12227     359  \n",
       "1       12721     109  \n",
       "2       14409    1189  \n",
       "3       12057      48  \n",
       "4       12227      49  \n",
       "...       ...     ...  \n",
       "771718  14673     525  \n",
       "771719  14742      69  \n",
       "771720  15887    1141  \n",
       "771721  15953      66  \n",
       "771722  16023      68  \n",
       "\n",
       "[771723 rows x 7 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc657674",
   "metadata": {},
   "source": [
    "## Adding Features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d32cc3f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gets position, type, and length info for genes and transcripts\n",
    "genes['gene_type'] = genes['gene_type'].map(lambda s: ' '.join(s.split('_')).capitalize())\n",
    "gene_len = genes[['gene_id', 'start', 'gene_type', 'gene_name']].rename(columns={'start': 'gene_start'})\n",
    "tx_len = transcripts[['transcript_id', 'length']].rename(columns={'length': 'tx_length'})\n",
    "exons = (exons\n",
    "         .merge(gene_len, on='gene_id')\n",
    "         .merge(tx_len, on='transcript_id')\n",
    "         [['chr', 'gene_id', 'gene_name', 'gene_type', 'gene_start',\n",
    "           'transcript_id', 'tx_length', 'exon_id', 'start', 'end', 'length']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e58bf8fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tx_length</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2541</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2541</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2541</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>1189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>1661</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>1661</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771718</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Protein coding</td>\n",
       "      <td>14149</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>525</td>\n",
       "      <td>ENSE00001434974.2</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771719</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Mt trna</td>\n",
       "      <td>14674</td>\n",
       "      <td>ENST00000387459.1</td>\n",
       "      <td>69</td>\n",
       "      <td>ENSE00001544476.1</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771720</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Protein coding</td>\n",
       "      <td>14747</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>1141</td>\n",
       "      <td>ENSE00001436074.2</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771721</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Mt trna</td>\n",
       "      <td>15888</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>66</td>\n",
       "      <td>ENSE00001544475.2</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771722</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt trna</td>\n",
       "      <td>15956</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>68</td>\n",
       "      <td>ENSE00001544473.2</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>771723 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr            gene_id gene_name                           gene_type  \\\n",
       "0       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "1       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "2       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "3       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "4       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "...      ...                ...       ...                                 ...   \n",
       "771718  chrM  ENSG00000198695.2    MT-ND6                      Protein coding   \n",
       "771719  chrM  ENSG00000210194.1     MT-TE                             Mt trna   \n",
       "771720  chrM  ENSG00000198727.2    MT-CYB                      Protein coding   \n",
       "771721  chrM  ENSG00000210195.2     MT-TT                             Mt trna   \n",
       "771722  chrM  ENSG00000210196.2     MT-TP                             Mt trna   \n",
       "\n",
       "        gene_start      transcript_id  tx_length            exon_id  start  \\\n",
       "0            11869  ENST00000456328.2       2541  ENSE00002234944.1  11869   \n",
       "1            11869  ENST00000456328.2       2541  ENSE00003582793.1  12613   \n",
       "2            11869  ENST00000456328.2       2541  ENSE00002312635.1  13221   \n",
       "3            11869  ENST00000450305.2       1661  ENSE00001948541.1  12010   \n",
       "4            11869  ENST00000450305.2       1661  ENSE00001671638.2  12179   \n",
       "...            ...                ...        ...                ...    ...   \n",
       "771718       14149  ENST00000361681.2        525  ENSE00001434974.2  14149   \n",
       "771719       14674  ENST00000387459.1         69  ENSE00001544476.1  14674   \n",
       "771720       14747  ENST00000361789.2       1141  ENSE00001436074.2  14747   \n",
       "771721       15888  ENST00000387460.2         66  ENSE00001544475.2  15888   \n",
       "771722       15956  ENST00000387461.2         68  ENSE00001544473.2  15956   \n",
       "\n",
       "          end  length  \n",
       "0       12227     359  \n",
       "1       12721     109  \n",
       "2       14409    1189  \n",
       "3       12057      48  \n",
       "4       12227      49  \n",
       "...       ...     ...  \n",
       "771718  14673     525  \n",
       "771719  14742      69  \n",
       "771720  15887    1141  \n",
       "771721  15953      66  \n",
       "771722  16023      68  \n",
       "\n",
       "[771723 rows x 11 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1d9565b",
   "metadata": {},
   "source": [
    "## Compute Lengths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "19b90a12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_bp(row):\n",
    "    starts = row['start']\n",
    "    ends = row['end']\n",
    "    start = min(starts)\n",
    "    end = max(ends)\n",
    "    bp = [False] * (end - start + 1)\n",
    "    for i in range(len(starts)):\n",
    "        s = starts[i] - start\n",
    "        e = ends[i] - start + 1\n",
    "        bp[s:e] = [True] * (e - s)\n",
    "    return sum(bp)\n",
    "\n",
    "# Computes length of merged transcripts\n",
    "merged_len = (exons\n",
    "              .groupby(['gene_id'])[['start', 'end']]\n",
    "              .agg(list)\n",
    "              .apply(count_bp, axis=1)\n",
    "              .to_frame('merged_length')\n",
    "              .reset_index())\n",
    "\n",
    "# Computes length of each transcript (sum of exons)\n",
    "transcript_len = exons.groupby(['gene_id', 'transcript_id'])['length'].sum()\n",
    "# Computes average, maximum, and median lengths over all transcripts\n",
    "avg_len = transcript_len.groupby(level=[0]).agg('mean').to_frame('mean_length').reset_index()\n",
    "max_len = transcript_len.groupby(level=[0]).agg('max').to_frame('max_length').reset_index()\n",
    "median_len = transcript_len.groupby(level=[0]).agg('median').to_frame('median_length').reset_index()\n",
    "\n",
    "# Computes number of transcripts per gene\n",
    "n_transcripts = (exons[['gene_id', 'transcript_id']]\n",
    "                 .drop_duplicates()\n",
    "                 .groupby(['gene_id'])['transcript_id']\n",
    "                 .count()\n",
    "                 .to_frame('n_tx')\n",
    "                 .reset_index())\n",
    "\n",
    "exons = (exons\n",
    "         .merge(n_transcripts, on='gene_id')\n",
    "         .merge(merged_len, on='gene_id')\n",
    "         .merge(avg_len, on='gene_id')\n",
    "         .merge(median_len, on='gene_id')\n",
    "         .merge(max_len, on='gene_id'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "16402974",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>tx_length</th>\n",
       "      <th>exon_id</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>n_tx</th>\n",
       "      <th>merged_length</th>\n",
       "      <th>mean_length</th>\n",
       "      <th>median_length</th>\n",
       "      <th>max_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2541</td>\n",
       "      <td>ENSE00002234944.1</td>\n",
       "      <td>11869</td>\n",
       "      <td>12227</td>\n",
       "      <td>359</td>\n",
       "      <td>2</td>\n",
       "      <td>1735</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2541</td>\n",
       "      <td>ENSE00003582793.1</td>\n",
       "      <td>12613</td>\n",
       "      <td>12721</td>\n",
       "      <td>109</td>\n",
       "      <td>2</td>\n",
       "      <td>1735</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>2541</td>\n",
       "      <td>ENSE00002312635.1</td>\n",
       "      <td>13221</td>\n",
       "      <td>14409</td>\n",
       "      <td>1189</td>\n",
       "      <td>2</td>\n",
       "      <td>1735</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>1661</td>\n",
       "      <td>ENSE00001948541.1</td>\n",
       "      <td>12010</td>\n",
       "      <td>12057</td>\n",
       "      <td>48</td>\n",
       "      <td>2</td>\n",
       "      <td>1735</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>1661</td>\n",
       "      <td>ENSE00001671638.2</td>\n",
       "      <td>12179</td>\n",
       "      <td>12227</td>\n",
       "      <td>49</td>\n",
       "      <td>2</td>\n",
       "      <td>1735</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1144.5</td>\n",
       "      <td>1657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771718</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Protein coding</td>\n",
       "      <td>14149</td>\n",
       "      <td>ENST00000361681.2</td>\n",
       "      <td>525</td>\n",
       "      <td>ENSE00001434974.2</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>525</td>\n",
       "      <td>1</td>\n",
       "      <td>525</td>\n",
       "      <td>525.0</td>\n",
       "      <td>525.0</td>\n",
       "      <td>525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771719</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Mt trna</td>\n",
       "      <td>14674</td>\n",
       "      <td>ENST00000387459.1</td>\n",
       "      <td>69</td>\n",
       "      <td>ENSE00001544476.1</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>69</td>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>69.0</td>\n",
       "      <td>69.0</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771720</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Protein coding</td>\n",
       "      <td>14747</td>\n",
       "      <td>ENST00000361789.2</td>\n",
       "      <td>1141</td>\n",
       "      <td>ENSE00001436074.2</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>1141</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>1141.0</td>\n",
       "      <td>1141.0</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771721</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Mt trna</td>\n",
       "      <td>15888</td>\n",
       "      <td>ENST00000387460.2</td>\n",
       "      <td>66</td>\n",
       "      <td>ENSE00001544475.2</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>66</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>66.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771722</th>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt trna</td>\n",
       "      <td>15956</td>\n",
       "      <td>ENST00000387461.2</td>\n",
       "      <td>68</td>\n",
       "      <td>ENSE00001544473.2</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>68</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>68.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>771723 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         chr            gene_id gene_name                           gene_type  \\\n",
       "0       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "1       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "2       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "3       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "4       chr1  ENSG00000223972.5   DDX11L1  Transcribed unprocessed pseudogene   \n",
       "...      ...                ...       ...                                 ...   \n",
       "771718  chrM  ENSG00000198695.2    MT-ND6                      Protein coding   \n",
       "771719  chrM  ENSG00000210194.1     MT-TE                             Mt trna   \n",
       "771720  chrM  ENSG00000198727.2    MT-CYB                      Protein coding   \n",
       "771721  chrM  ENSG00000210195.2     MT-TT                             Mt trna   \n",
       "771722  chrM  ENSG00000210196.2     MT-TP                             Mt trna   \n",
       "\n",
       "        gene_start      transcript_id  tx_length            exon_id  start  \\\n",
       "0            11869  ENST00000456328.2       2541  ENSE00002234944.1  11869   \n",
       "1            11869  ENST00000456328.2       2541  ENSE00003582793.1  12613   \n",
       "2            11869  ENST00000456328.2       2541  ENSE00002312635.1  13221   \n",
       "3            11869  ENST00000450305.2       1661  ENSE00001948541.1  12010   \n",
       "4            11869  ENST00000450305.2       1661  ENSE00001671638.2  12179   \n",
       "...            ...                ...        ...                ...    ...   \n",
       "771718       14149  ENST00000361681.2        525  ENSE00001434974.2  14149   \n",
       "771719       14674  ENST00000387459.1         69  ENSE00001544476.1  14674   \n",
       "771720       14747  ENST00000361789.2       1141  ENSE00001436074.2  14747   \n",
       "771721       15888  ENST00000387460.2         66  ENSE00001544475.2  15888   \n",
       "771722       15956  ENST00000387461.2         68  ENSE00001544473.2  15956   \n",
       "\n",
       "          end  length  n_tx  merged_length  mean_length  median_length  \\\n",
       "0       12227     359     2           1735       1144.5         1144.5   \n",
       "1       12721     109     2           1735       1144.5         1144.5   \n",
       "2       14409    1189     2           1735       1144.5         1144.5   \n",
       "3       12057      48     2           1735       1144.5         1144.5   \n",
       "4       12227      49     2           1735       1144.5         1144.5   \n",
       "...       ...     ...   ...            ...          ...            ...   \n",
       "771718  14673     525     1            525        525.0          525.0   \n",
       "771719  14742      69     1             69         69.0           69.0   \n",
       "771720  15887    1141     1           1141       1141.0         1141.0   \n",
       "771721  15953      66     1             66         66.0           66.0   \n",
       "771722  16023      68     1             68         68.0           68.0   \n",
       "\n",
       "        max_length  \n",
       "0             1657  \n",
       "1             1657  \n",
       "2             1657  \n",
       "3             1657  \n",
       "4             1657  \n",
       "...            ...  \n",
       "771718         525  \n",
       "771719          69  \n",
       "771720        1141  \n",
       "771721          66  \n",
       "771722          68  \n",
       "\n",
       "[771723 rows x 16 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exons"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dadf3af3",
   "metadata": {},
   "source": [
    "## Assembling and Saving Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e9e03031",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assembles final table\n",
    "result = exons[['chr', 'gene_id', 'gene_name', 'gene_type', 'gene_start', 'n_tx', 'merged_length', 'mean_length', 'median_length', 'max_length']].drop_duplicates()\n",
    "# Assigns sequential gene number within chromosome\n",
    "result['seq'] = result.assign(seq=1).groupby(['chr'])['seq'].agg('cumsum')\n",
    "# Turns mean and median lengths into integers\n",
    "result['mean_length'] = result['mean_length'].astype(int)\n",
    "result['median_length'] = result['median_length'].astype(int)\n",
    "\n",
    "# Selects the columns in the right order\n",
    "result = result[['seq', 'chr', 'gene_id', 'gene_name', 'gene_type', 'gene_start', 'n_tx', 'merged_length', 'mean_length', 'median_length', 'max_length']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "34486f16",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>chr</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>n_tx</th>\n",
       "      <th>merged_length</th>\n",
       "      <th>mean_length</th>\n",
       "      <th>median_length</th>\n",
       "      <th>max_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000223972.5</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>Transcribed unprocessed pseudogene</td>\n",
       "      <td>11869</td>\n",
       "      <td>2</td>\n",
       "      <td>1735</td>\n",
       "      <td>1144</td>\n",
       "      <td>1144</td>\n",
       "      <td>1657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000227232.5</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>Unprocessed pseudogene</td>\n",
       "      <td>14404</td>\n",
       "      <td>1</td>\n",
       "      <td>1351</td>\n",
       "      <td>1351</td>\n",
       "      <td>1351</td>\n",
       "      <td>1351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000278267.1</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>Mirna</td>\n",
       "      <td>17369</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000243485.5</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>Lncrna</td>\n",
       "      <td>29554</td>\n",
       "      <td>2</td>\n",
       "      <td>1021</td>\n",
       "      <td>623</td>\n",
       "      <td>623</td>\n",
       "      <td>712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>ENSG00000284332.1</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>Mirna</td>\n",
       "      <td>30366</td>\n",
       "      <td>1</td>\n",
       "      <td>138</td>\n",
       "      <td>138</td>\n",
       "      <td>138</td>\n",
       "      <td>138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771718</th>\n",
       "      <td>33</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000198695.2</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>Protein coding</td>\n",
       "      <td>14149</td>\n",
       "      <td>1</td>\n",
       "      <td>525</td>\n",
       "      <td>525</td>\n",
       "      <td>525</td>\n",
       "      <td>525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771719</th>\n",
       "      <td>34</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210194.1</td>\n",
       "      <td>MT-TE</td>\n",
       "      <td>Mt trna</td>\n",
       "      <td>14674</td>\n",
       "      <td>1</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771720</th>\n",
       "      <td>35</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000198727.2</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>Protein coding</td>\n",
       "      <td>14747</td>\n",
       "      <td>1</td>\n",
       "      <td>1141</td>\n",
       "      <td>1141</td>\n",
       "      <td>1141</td>\n",
       "      <td>1141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771721</th>\n",
       "      <td>36</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210195.2</td>\n",
       "      <td>MT-TT</td>\n",
       "      <td>Mt trna</td>\n",
       "      <td>15888</td>\n",
       "      <td>1</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>771722</th>\n",
       "      <td>37</td>\n",
       "      <td>chrM</td>\n",
       "      <td>ENSG00000210196.2</td>\n",
       "      <td>MT-TP</td>\n",
       "      <td>Mt trna</td>\n",
       "      <td>15956</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60660 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seq   chr            gene_id    gene_name  \\\n",
       "0         1  chr1  ENSG00000223972.5      DDX11L1   \n",
       "9         2  chr1  ENSG00000227232.5       WASH7P   \n",
       "20        3  chr1  ENSG00000278267.1    MIR6859-1   \n",
       "21        4  chr1  ENSG00000243485.5  MIR1302-2HG   \n",
       "26        5  chr1  ENSG00000284332.1    MIR1302-2   \n",
       "...     ...   ...                ...          ...   \n",
       "771718   33  chrM  ENSG00000198695.2       MT-ND6   \n",
       "771719   34  chrM  ENSG00000210194.1        MT-TE   \n",
       "771720   35  chrM  ENSG00000198727.2       MT-CYB   \n",
       "771721   36  chrM  ENSG00000210195.2        MT-TT   \n",
       "771722   37  chrM  ENSG00000210196.2        MT-TP   \n",
       "\n",
       "                                 gene_type  gene_start  n_tx  merged_length  \\\n",
       "0       Transcribed unprocessed pseudogene       11869     2           1735   \n",
       "9                   Unprocessed pseudogene       14404     1           1351   \n",
       "20                                   Mirna       17369     1             68   \n",
       "21                                  Lncrna       29554     2           1021   \n",
       "26                                   Mirna       30366     1            138   \n",
       "...                                    ...         ...   ...            ...   \n",
       "771718                      Protein coding       14149     1            525   \n",
       "771719                             Mt trna       14674     1             69   \n",
       "771720                      Protein coding       14747     1           1141   \n",
       "771721                             Mt trna       15888     1             66   \n",
       "771722                             Mt trna       15956     1             68   \n",
       "\n",
       "        mean_length  median_length  max_length  \n",
       "0              1144           1144        1657  \n",
       "9              1351           1351        1351  \n",
       "20               68             68          68  \n",
       "21              623            623         712  \n",
       "26              138            138         138  \n",
       "...             ...            ...         ...  \n",
       "771718          525            525         525  \n",
       "771719           69             69          69  \n",
       "771720         1141           1141        1141  \n",
       "771721           66             66          66  \n",
       "771722           68             68          68  \n",
       "\n",
       "[60660 rows x 11 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86b7ecc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "result.to_csv('genes.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
